### ISSUE TITLE
A single, public, Python API for installing package

### ISSUE THREAD

**Issue by ryanking13** [pyodide/pyodide#4914]
### Proposal

I would like to make a single, public, Python, API for installing a package. Here, "install" does not involve loading the package from an external URL or doing dependency resolution. Instead, it means unpacking the package into the proper directory and doing some post installation steps.

This API will be used inside `pyodide.loadPackage,` `micropip.install`, and can be used in other downstream projects as well.

Currently, we are using [`pyodide._package_loader.unpack_buffer`](https://github.com/pyodide/pyodide/blob/8ba8f4497cf25d6ec876fb2703dc77afa5f93bae/src/py/pyodide/_package_loader.py#L143) for a similar purpose, but it has following limitations.

1) It is a private API
2) It does not cover loading shared libraries
3) It's name isn't that straightforward.

So I would like to make a new API that would have signature such as:

```python
async install_package(
  buffer: bytes | JsBuffer,
  format: ...
  extract_dir: ...
  load_dynlibs: ...
  load_datafiles: ...
  <something else>
):
  # handles installing a package
  # A package can be a
  # 1. Python wheel
  # 2. Other pyodide-supported types (shared library in a .zip format)
  # 3. tar or zip archive?

  # This function should also handle
  # - loading shared libraries
  # - installing data_files
  # - etc
```

### Motivation

I would like to make `pyodide.loadPackage` and `micropip.install` use the same API when installing a package. For now, they are not shared very well. Therefore, for instance, if I want to add `data_files` support in both `pyodide.loadPackage` and `micropip.install`, I need to implement it separately in `pyodide._package_loader` and in `micropip.WheelInfo`, which is quite annoying.

**Comment by hoodmane** [comment]
Sounds good to me!

**Comment by hoodmane** [comment]
For stack switching, it would be nice if we could implement this in JavaScript. I think it's possible with `decompressionStream` at this point.

**Comment by ryanking13** [comment]
> For stack switching, it would be nice if we could implement this in JavaScript. I think it's possible with decompressionStream at this point.

I see. Changing the top-level entry point to JavaScript seems fine to me. It'll probably use information from both Python and JS internally anyway

**Comment by ShaheedHaque** [comment]
Worth considering #4596?

**Comment by hoodmane** [comment]
> It'll probably use information from both Python and JS internally anyway

Ideally for memory snapshots I think it should be JS only. Though perhaps this means that I should be snapshoting the entire file system so that the package loader is not needed at all to initialize packages.

**Comment by ryanking13** [comment]
@ShaheedHaque

> Worth considering https://github.com/pyodide/pyodide/issues/4596?

I think that issue is not that related to this. But it's something we need to do anyway.

**Comment by ryanking13** [comment]
> Ideally for memory snapshots I think it should be JS only. Though perhaps this means that I should be snapshoting the entire file system so that the package loader is not needed at all to initialize packages.

Is it because loading the memory snapshot should run before installing the Python codes? But is memory snapshot something to do with installing a package?

**Comment by hoodmane** [comment]
The problem is with shared library loading. The shared libraries and their symbols are present in C memory and they depend on JavaScript runtime state to work. If we don't load the same set of shared libs and allocate their metadata and handles in the same memory locations, then calling the symbols that were dlsym'd from them will segfault. But then this means that we need to load any packages prior to restoring the snapshot. But the package loader depends on Python so that's problematic because we'd rather not use Python at all until we restore the snapshot (otherwise we end up doing some work twice).

Probably the best solution though is to serialize the entire file system state. If we don't do that, we'll need to remove all usage of Python from `pyodide.loadPackage`.

**Comment by ryanking13** [comment]
> Probably the best solution though is to serialize the entire file system state. If we don't do that, we'll need to remove all usage of Python from pyodide.loadPackage.

Yeah, I thought that memory snapshot would work by capturing the entire file system state in the end. Anyway, I think making the package loader not to rely on Python as much as possible is a thing that we can try.
